doc: add extension for external content

gmarull opened the pull request:
The `external_content` extension allows gathering sources from multiple locations. It replaces the `extract_content` script, which was invoked by CMake.

Sphinx defaults have been adjusted to offer better user experience (i.e. interactive and colored output by default).

This change also fixes an old issue with current script: deleted files were not tracked, so they generated Sphinx build errors on the next build.

gmarull committed:
doc: improve Sphinx defaults

Make default Sphinx options more user-friendly, i.e. colored output so
that one can easily follow build progress.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: extensions: initial version of external_content

external_content extension allows to gather content from arbitrary
locations.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: use external_content

Use external_content Sphinx extensions to gather documentation sources.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: scripts: remove extract_content

Remove unused extract_content script.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

mbolivar-nordic commented on doc/CMakeLists.txt:
```diff
@@ -59,7 +59,7 @@ if(${LATEXMK} STREQUAL LATEXMK-NOTFOUND)
 
 if(NOT DEFINED SPHINXOPTS)
-  set(SPHINXOPTS -q -j auto)
+  set(SPHINXOPTS -j auto)
```
This is making the output for the documentation build step in CI pretty useless; can you tweak this so that is done the same way, with -q etc.?

gmarull replied:
I'll add `-q` to CI

gmarull committed:
ci: doc: force quiet mode

Force quiet mode in CI, since interactive output adds no value there.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

mbolivar-nordic commented on .github/workflows/doc-build.yml:
```diff
@@ -56,7 +56,7 @@ jobs:
       run: |
         source zephyr-env.sh
-        make -C doc htmldocs
+        SPHINXOPTS="'-q -j auto'" make -C doc htmldocs
```
Do you need all these quotes? Wouldn't `SPHINXOPTS="-q -jauto"` be enough?

mbolivar-nordic replied:
What about colored output; don't you want to disable it here too?

gmarull replied:
Yes, otherwise you get:
```
cmake -GNinja -B_build -S. -DDOC_TAG=development \
        -DSPHINXOPTS=-q -j auto \
        -DKCONFIG_TURBO_MODE=0
CMake Error: Unknown argument -j
```
instead of
```
cmake -GNinja -B_build -S. -DDOC_TAG=development \
        -DSPHINXOPTS='-q -j auto' \
        -DKCONFIG_TURBO_MODE=0
```
any more elegant suggestion?

gmarull replied:
> What about colored output; don't you want to disable it here too?

Hmm according to help, it auto-detects color availability: `  -N, --no-color    do not emit colored output (default: auto-detect)`, I imagine colored output will not be available on CI.

mbolivar-nordic replied:
> any more elegant suggestion?

That looks like papering over a bug in the makefile; it should be:

```patch
diff --git a/doc/Makefile b/doc/Makefile
index 12981e465c..08cae1f870 100644
--- a/doc/Makefile
+++ b/doc/Makefile
@@ -23,7 +23,7 @@ htmldocs pdfdocs doxygen: configure
 .PHONY: configure
 configure:
        cmake -GNinja -B${BUILDDIR} -S. -DDOC_TAG=${DOC_TAG} \
-               -DSPHINXOPTS=${SPHINXOPTS} \
+               -DSPHINXOPTS="${SPHINXOPTS}" \
                -DKCONFIG_TURBO_MODE=${KCONFIG_TURBO_MODE}
 
 clean:
```

gmarull replied:
good catch, fixed

gmarull committed:
doc: makefile: adjust default Sphinx options

Adjust default options to match the ones in CMake. Quote options when
invoking CMake.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
latest push: use Sphinx `source_encoding` when reading files.

gmarull commented:
latest push: take into account deleted files so that incremental builds work even when deleting files
